src/net/client/cache.rs

```diff
@@ -0,0 +1,1177 @@
+    }
+
+    let mut msg = msg.next_section()?.expect("section should be present");
+    for rr in &mut msg {
+        let rr = rr?;
+        min_val =
+            min(min_val, Duration::from_secs(rr.ttl().as_secs() as u64));
+    }
+
+    let msg = msg.next_section()?.expect("section should be present");
+    for rr in msg {
+        let rr = rr?;
+        if rr.rtype() != Rtype::Opt {
+            min_val =
+                min(min_val, Duration::from_secs(rr.ttl().as_secs() as u64));
+        }
+    }
+
+    Ok(min_val)
+}
+
+/// Return a new message with decremented TTL values.
+fn decrement_ttl<TDN>(
+    orig_qname: TDN,
+    response: &Result<Message<Bytes>, Error>,
+    amount: u32,
+) -> Result<Message<Bytes>, Error>
+where
+    TDN: ToDname + Clone,
+{
+    let msg = match response {
+        Err(err) => return Err(err.clone()),
+        Ok(msg) => msg,
+    };
+
+    let amount = Ttl::from_secs(amount);
+
+    let mut target =
+        MessageBuilder::from_target(StaticCompressor::new(Vec::new()))
+            .expect("Vec is expected to have enough space");
+
+    let source = msg;
+
+    *target.header_mut() = source.header();
+
+    let source = source.question();
+    let mut target = target.question();
+    for rr in source {
+        let rr = rr?;
+        target
+            .push((orig_qname.clone(), rr.qtype(), rr.qclass()))
+            .expect("push failed");
+    }
+    let mut source = source.answer()?;
+    let mut target = target.answer();
+    for rr in &mut source {
+        let mut rr = rr?
+            .into_record::<AllRecordData<_, ParsedDname<_>>>()?
+            .expect("record expected");
+        rr.set_ttl(rr.ttl() - amount);
+        target.push(rr).expect("push failed");
+    }
+
+    let mut source =
+        source.next_section()?.expect("section should be present");
+    let mut target = target.authority();
+    for rr in &mut source {
+        let mut rr = rr?
+            .into_record::<AllRecordData<_, ParsedDname<_>>>()?
+            .expect("record expected");
+        rr.set_ttl(rr.ttl() - amount);
+        target.push(rr).expect("push failed");
+    }
+
+    let source = source.next_section()?.expect("section should be present");
+    let mut target = target.additional();
+    for rr in source {
+        let rr = rr?;
+        let mut rr = rr
+            .into_record::<AllRecordData<_, ParsedDname<_>>>()?
+            .expect("record expected");
+        if rr.rtype() != Rtype::Opt {
+            rr.set_ttl(rr.ttl() - amount);
+        }
+        target.push(rr).expect("push failed");
+    }
+
+    let result = target.as_builder().clone();
+    let msg =
+        Message::<Bytes>::from_octets(result.finish().into_target().into())
+            .expect(
+                "Message should be able to parse output from MessageBuilder",
+            );
+    Ok(msg)
+}
+
+/// Return a new message without the DNSSEC type RRSIG, NSEC, and NSEC3.
+fn remove_dnssec(
```

Comment:
Could this be a convenience method on `Message`?

Reply:
I'll leave this in cache at least until it is clear what the validator needs. There are three problems: the name. It removes only some DNSSEC records. It can clear the AD flag, not clear if that is needed in general (or that it should do more, may something for plain), It removes records even if they are requested by the question.